### DISCUSSION TITLE
Type issues when using `createStore` in TS

### DISCUSSION THREAD

**Discussion by motistudio** [pmndrs/jotai#2837] [General]
Hi

I've noticed a type issue when using `createStore` in a TS environment.
Whenever it's used, the following error appears:
```
Exported variable 'store' has or is using name 'AtomState' from external module ".../node_modules/jotai/esm/vanilla/store" but cannot be named
```

### TS config
I think it can be fixed by tweeking the `module` & `moduleResolution`, but that's not possible for every project.
In my case it's:
```json
{
  "compilerOptions": {
    "useDefineForClassFields": true,
    "lib": ["ES2020", "DOM", "DOM.Iterable"],
    "skipLibCheck": true,
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "allowSyntheticDefaultImports": true,
    "forceConsistentCasingInFileNames": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "strictNullChecks": true,
    "jsx": "react-jsx",
    "strict": true,
  }
}
```

### Solution
The solution is to just export the `Store` type, allowing the type "to be named".
I think it's useful anyhow, since users who use `createStore` might need to describe it as a type as well.

A wider solution would be to not use the `export ... from '...'` syntax, it seems that it's the primary case of this specific error across different packages.

--

Thank you!
Loves this library

**Comment by dai-shi** [comment]
Thanks for reporting.
We had a similar discussion before and our workaround for now is to explicitly type with `ReturnType<typeof createStore>` on the user end. If you need help, please create a reproduction with TypeScript Playground.
For now, the ways we limit exporting types is intentional. I think the workaround should probably be documented.

**Reply by uxduck** [reply]
I've been using `export type JotaiStore = ReturnType<typeof createStore>;` successfully for more than a year, but updating to the latest jotai, I now get a TS error:

```
src/routes/__root.tsx:30:19 - error TS2345: Argument of type 'PrdOrDevStore' is not assignable to parameter of type 'PrdOrDevStore'.
  Type 'Store' is not assignable to type 'PrdOrDevStore'.
    Type 'Store' is not assignable to type 'Store & DevStoreRev4'.
      Property 'unstable_derive' is missing in type 'Store' but required in type 'Store'.

30   useThemeDomSync(jotaiStore);
```
```
export type JotaiStore = ReturnType<typeof createStore>;

// Syncs the DOM theme (dark class) with the store theme
export const useThemeDomSync = (store: JotaiStore) => {
```

The cause was a mismatch of versions (2 versions) of jotai. I just need to make sure i use the same jotai version

**Reply by dai-shi** [reply]
Nice catch. Yeah, the types are slightly changed in the recent versions. This is for migration at the moment and the future versions will probably resolve it.
